htdocs/adherents/class/adherentstats.class.php

```diff
@@ -313,16 +315,68 @@ public function countMembersByTagAndStatus($numberYears = 0)
 				foreach ($MembersCountArray[$objp->fk_categorie] as $key=>$nb) {
 					if ($key!='label') {
 						$totalrow += $nb;
-						$totalstatus[$key] += $nb;
+						// total row sum only root tag cause all sublevel are already included in root level
+						if ($MembersCountArray[$objp->fk_categorie]['fk_parent'] == 0) {
+							$totalstatus[$key] += $nb;
+						}
 					}
 				}
 				$MembersCountArray[$objp->fk_categorie]['total_adhtag'] = $totalrow;
 				$i++;
 			}
 			$this->db->free($result);
+			$MembersCountArray = $this->buildTree($MembersCountArray, 0);
 			$MembersCountArray['total'] = $totalstatus;
 			$MembersCountArray['total']['all'] = array_sum($totalstatus);
+			$MembersCountArray['arraydepth'] = $this->arrayDepth($MembersCountArray);
 		}
 		return $MembersCountArray;
 	}
+
+	/**
+	 *	Recursive function returning a multidimensional array representing
+	 *  a parent-child tree with the addition of a "depth" property at each level
+	 *
+	 * @param		array	$elements 		Array of elements in wich build a tree
+	 * @param		int		$fk_parent		First parent branch to build the tree (0 is the trunk for tags)
+	 * @param		int		$limit_depth    Maximum depth, beyond which elements are deleted from the returned tree
+	 * @return		array 					Multidimensional array representing a tree with a "depth" property
+	 */
+	public function buildTree(array $elements, int $fk_parent = 0, int $limit_depth = -1)
+	{
+		$branch = array();
+		$depth = $limit_depth > 0 ? $limit_depth : -($limit_depth + 1);
+
+		foreach ($elements as $key=>$element) {
+			if (is_array($element)) {
+				$element['depth'] = $depth;
+			}
+			if (isset($element['fk_parent']) && $element['fk_parent'] == $fk_parent) {
+				$children = ($limit_depth != 0) ? $this->buildTree($elements, $key, $limit_depth-1) : array();
+				if ($children) {
+					$element['children'] = $children;
+				}
+				$branch[] = $element;
+			}
+		}
+		return $branch;
+	}
+
+	/**
+	 *	Recursive function returning the deepest level of a multidimentionnal array
+	 *
+	 * @param		array	$array 		Array of elements in wich build a tree
+	 * @return		int 	Maximum tree depth
+	 */
+	public function arrayDepth(array $array)
+	{
+		$max_depth = 0;
+		foreach ($array as $value) {
+			if (is_array($value)) {
+				$depth = array_key_exists('label', $value) ? $this->arrayDepth($value) + 1 : $this->arrayDepth($value);
```

Comment:
Because this is a recursive method, please add a test protection
`if $depth > 1000 then exit `   (should not happen, but happen when data are corrupted for example after direct database updates).